src/anet.c

```diff
@@ -591,7 +610,10 @@ static int _anetTcpServer(char *err, int port, char *bindaddr, int af, int backl
         return ANET_ERR;
     }
     for (p = servinfo; p != NULL; p = p->ai_next) {
-        if ((s = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) continue;
+        rv = anetTcpGetProtocol(err, p->ai_protocol, mptcp);
+        if (rv == ANET_ERR) goto error;
+
+        if ((s = socket(p->ai_family, p->ai_socktype, rv)) == -1) continue;
```

Comment:
@pizhenwei if Valkey maintainers are still OK with that, I think it would make sense to create an MPTCP socket by default, if available. The option can still be used to turn it off if really needed. Then, it will be up to the client to decide to use MPTCP or not: that will ease the deployments. If MPTCP is only needed for some specific clients, then they can use it, and the others can continue to use "plain" TCP like before.

So here, `anetTcpGetProtocol` can be dropped, and something like this can be done:

```suggestion
#ifdef IPPROTO_MPTCP
        if (!mptcp || (s = socket(p->ai_family, p->ai_socktype, IPPROTO_MPTCP)) == -1)
#endif
        {
                if ((s = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) == -1) continue;
        }
```

WDYT?

Reply:
Hi @matttbe 

* explicit mptcp (not strong opinion): I still prefer explicit mptcp config for the server side, because TCP and MPTCP are different transmission protocol theoretically. The maintainers (who deploy valkey-server) of valkey services should know the transmission protocol of their service and necessary operations for trouble shooting.
* default no (not strong opinion): coping the config from 8.0 for the new version, but it listens on different socket protocol, it will make user confused.
* `anetTcpGetProtocol` is needed (strong opinion): if user specifies `mptcp yes` on a lower kernel or other OS, it should be an error, because the valkey-server could NOT satisfy the user's request.

Let's leave this to the maintainer to decide...

Reply:
@pizhenwei thank you for your reply.

> explicit mptcp (not strong opinion): I still prefer explicit mptcp config for the server side, because TCP and MPTCP are different transmission protocol theoretically. The maintainers (who deploy valkey-server) of valkey services should know the transmission protocol of their service and necessary operations for trouble shooting.

MPTCP is an extension to TCP, it is not a radical change. On the wire, TCP is still used, but with extra options in the header. But indeed, sysadmins should be told about the behaviour change, but at the end, they should not worry about that. Please note that when the client doesn't request to use MPTCP, the server application will get a "plain" TCP socket after the `accept()`. So no modification on that side.

> * `anetTcpGetProtocol` is needed (strong opinion): if user specifies `mptcp yes` on a lower kernel or other OS, it should be an error, because the valkey-server could NOT satisfy the user's request.

If MPTCP is used by default, the application should not fail if it is not available. Or you need tree choices: auto (fallback to TCP if MPTCP is not available), on (error if not available), off (don't try to use MPTCP).

> Let's leave this to the maintainer to decide...

Indeed, my proposal was just a suggestion to ease MPTCP deployments without extra steps on the server side, only on the client side.

Reply:
I understand it would be nice to enable it by default so the client can make the decisions. I also read https://www.mptcp.dev/faq.html#why--when-should-mptcp-be-enabled-by-default.

But I think we should have it disabled by default, for the following reasons:

* Uniform experience. On other platforms, like BSD and on older Linux, don't have this MPTCP. If a user tries to enable it where it's not supported, I like an error instead of a silent fallback.
* It's a new feature. We often add new features disabled by default so users can opt in to try them. In a future version, we can consider making it enabled (or "auto") by default.

Reply:
> I understand it would be nice to enable it by default so the client can make the decisions. I also read https://www.mptcp.dev/faq.html#why--when-should-mptcp-be-enabled-by-default.
> 
> But I think we should have it disabled by default, for the following reasons:
> 
> * Uniform experience. On other platforms, like BSD and on older Linux, don't have this MPTCP. If a user tries to enable it where it's not supported, I like an error instead of a silent fallback.

Right, if a user specifies the MPTCP config on an unsupported platform, an explicit error is better than a silent fallback.

> * It's a new feature. We often add new features disabled by default so users can opt in to try them. In a future version, we can consider making it enabled (or "auto") by default.

Yes, MPTCP is a recent feature.

Fully agree!